heuristic/heuristic_calculations.py

```diff
@@ -1,322 +1,211 @@
+        #     print("Different isolated off days")
     return isolated_off_days
 
 
-def calculate_consecutive_days(data, x):
-
-    shifts_per_day = data["shifts"]["shifts_per_day"]
+def calculate_consecutive_days(model, x):
     consecutive_days = {}
-
-    for e in data["staff"]["employees"]:
-        for i in range(len(data["time"]["days"]) - data["limit_on_consecutive_days"]):
-
-            consecutive_days[e, i] = max(
-                0,
-                (
-                    sum(
-                        sum(x[e, t, v] for t, v in shifts_per_day[i_marked])
-                        for i_marked in range(i, i + data["limit_on_consecutive_days"])
-                    )
-                )
-                - data["limit_on_consecutive_days"],
-            )
+    for e in model.employees:
+        for i in range(len(model.days)-model.L_C_D):
+            consecutive_days[e,i] = max(0,(sum(
+                sum(x[e,t,v] for t,v in model.shifts_at_day[i_marked]) 
+            for i_marked in range(i,i+model.L_C_D)))- model.L_C_D)
+            
+            # if(consecutive_days[e,i] != model.q_con[e,i].x):
+            #     print("Different consecutive days")
     return consecutive_days
 
-
-def calculate_f(data, soft_vars, w, employees=None):
-
-    if not employees:
-        employees = data["staff"]["employees"]
-
+def calculate_f(model, soft_vars, w, employees=None):
+    if(employees == None):
+        employees = model.employees
     f = {}
-
-    # todo: add number_of_days to save calculations?
-    days = data["time"]["days"]
-
     for e in employees:
-        f[e] = (
-            sum(w[e, j][1] for j in data["time"]["weeks"])
-            - sum(soft_vars["deviation_contracted_hours"][e, j] for j in data["time"]["weeks"])
-            - sum(soft_vars["partial_weekends"][e, i] for i in data["time"]["saturdays"])
-            - sum(
-                soft_vars["isolated_working_days"][e, i + 1]
-                + soft_vars["isolated_off_days"][e, i + 1]
-                for i in range(len(days) - 2)
-            )
-            - sum(
-                soft_vars["consecutive_days"][e, i]
-                for i in range(len(days) - data["limit_on_consecutive_days"])
-            )
-        )
+        f[e] = (sum(w[e,j][1] for j in model.weeks)
+            - sum(soft_vars["contracted_hours"][e,j] for j in model.weeks)
+            - sum(soft_vars["partial_weekends"][e,i] for i in model.saturdays)
+            - sum(soft_vars["isolated_working_days"][e,i+1] + soft_vars["isolated_off_days"][e,i+1] for i in range(len(model.days)-2))
+            - sum(soft_vars["consecutive_days"][e,i] for i in range(len(model.days)-model.L_C_D)))
     return f
 
 
 def calculate_objective_function(model, soft_vars, w):
-
     f = calculate_f(model, soft_vars, w)
     g = min(f.values())
+    objective_function_value = (sum(f.values()) + g - abs(sum(soft_vars["deviation_from_ideal_demand"].values())))
```

Comment:
I think using a list for `f` could lead to significant speed improvements. What do you think?

Reply:
My reason for using dicts for most of the implementation is because I want it to be fool proof of the data we use. If we start using lists we should be certain that when indexing this list we know that we are using the correct values. If for example we don't have employees from 1 - 10, but perhaps 1, 4, 6, 8 ,10 then indexing by employee_id would not be possible. Since we use employee_id that starts at 1 and indexing starts at 0 we could easily get one-off-errors

Reply:
Ahh, good point! List is out of the picture then! Maybe a heapq with tuple-indices could be something, but only if we have time for it